### DISCUSSION TITLE
WCF - IBrowserProcess is null, unable to bind object

### DISCUSSION THREAD

**Discussion by Priyankay151** [cefsharp/CefSharp#4529] [General]
What version of the product are you using?

    112.3.0, installed via the Visual Studio solution nuget pkg manager

What architecture x86 or x64?

    x86

On what operating system?

    Win10

Are you using WinForms, WPF or OffScreen?

    WinForm

Issue 
While trying to bind object `ExecuteJavaScriptAsync("CefSharp.BindObjectAsync(\"" + <obje name> + "\")");` getting the   ERROR:JavascriptRootObjectWrapper.cpp(37)] IBrowserProcess is null, unable to bind object <obj name> . 

Debugging Observations - 
I am using CefSharp to render React widgets and HTML pages in my WPF application. Recently, I upgraded CefSharp version from v84.4.10 to v112.3.0. I have many flows where I'm using CEF to render the screens. 
while debugging I came across some observations :- 

1. I'm getting `ERROR:JavascriptRootObjectWrapper.cpp(37)] IBrowserProcess is null, unable to bind object <obj name>` only in one flow where I'm using `wcfenabled = true` and `async=false` while registering the object. In other flows it's `async=true` using IPC channel which are working fine. This issue coming in v112.3.0 not with v84.4.10.
2. If I first go to any other flow and then I go to that WCF enabled flow it gives above error but if I firstly go to that WCF enabled flow it works fine. Further I noticed that every time I create a new CefSharp WinForms control, two CefSharp.BrowserSubprocesses are getting created. While disposing that control, only one CefSharp.BrowserSubprocess is getting deleted, and the other one remains there. Now after disposing one instance I go to another flow then it will create only one CefSharp.browsesubprocess in case of remained CefSharp.browsersubprocess but if I delete that previous remained cef instace CefSharp.browsersubprocess from task manager and then I go to another flow then it will create two CefSharp.browsersubprocess. 
3. So in the case that remained CefSharp.browsersubprocess is not there, either firstly i go directly to that flow or before going to that flow I delete any remained CefSharp.browsesubprocess from task manager , in both cases that `wcfenabled = true` and `async=false` flow is working fine.

Questions - 

1. Why I'm getting ERROR:JavascriptRootObjectWrapper.cpp(37)] IBrowserProcess is null, unable to bind object <obj name>` , what are all the cases where I can get this error?
2. Why this error in flow where I'm using `wcfenabled = true` and `async=false` while registering the object ?
3. Why every cef control creating two browsersubprocess and while calling dispose() function only one is getting delete, how can i delete both? I can't use cef.shutdown() as I have many flows where CEF is used so cef.shutdown() I'm calling on exiting the application.
4. How can I fix this error? As I can't make it `async=true` in javascriptobjectrepo due to large code changes and that breaks the feature also in my application.

I looked into discussion #4519 and #915 . there's no other application running that uses WCF NetNamedPipeBinding.
I would appreciate any help or suggestions on how to resolve these issues with CefSharp 112.3.0 in my WPF application. Thank you.

**Comment by amaitland** [comment]
> With site-per-process enabled (default) a spare renderer process will be created for use with a future browser or navigation.

You can read more about the spare render process at https://github.com/chromiumembedded/cef/issues/2498#issuecomment-1465014269 and  https://github.com/chromiumembedded/cef/issues/2498#issuecomment-1465014266 

There's no `API` to close the `Spare Render Process`. 

You should be able to disable the spare render process using the following command line arg.

```
--disable-features=SpareRendererForSitePerProcess
```

> 2\. Further I noticed that every time I create a new CefSharp WinForms control, two CefSharp.BrowserSubprocesses are getting created. While disposing that control, only one CefSharp.BrowserSubprocess is getting deleted, and the other one remains there. Now after disposing one instance I go to another flow then it will create only one CefSharp.browsesubprocess in case of remained CefSharp.browsersubprocess but if I delete that previous remained cef instace CefSharp.browsersubprocess from task manager and then I go to another flow then it will create two CefSharp.browsersubprocess.

The `spare` render process is expected. It's not clear how this is related to the issue you are having. You can try disable the spare render process as above.

> 2\. If I first go to any other flow and then I go to that WCF enabled flow it gives above error but if I firstly go to that WCF enabled flow it works fine.

Make sure you set `CefSharpSettings.WcfEnabled = true` before creating any `ChromiumWebBrowser` instances. Past that I have no idea what your `flow` actually does, so it's impossible to say what's going on. See https://stackoverflow.com/help/minimal-reproducible-example for some guidance.

**Comment by Priyankay151** [comment]
Thanks @amaitland  Using '--disable-features=SpareRendererForSitePerProcess' worked for me. 
I have some followup questions over it - 

1. Can you please refer some wiki or docs to read more about spare render processs and '--disable-features=SpareRendererForSitePerProcess' ?
2. Will using the '--disable-features=SpareRendererForSitePerProcess' impact anything on the application like performance , cef rendering time etc. ?

**Reply by amaitland** [reply]
It sounds like the spare render process was launched with `WCF` disabled. Setting `CefSharpSettings.WcfEnabled = true` before calling `Cef.Initialze` should resolve the problem and is the approach I'd suggest. 

Best to keep with the default `Chromium` behaviour and leave the spare render process enabled.
